File: backend/legacy_ts_archive/src/services/AuthService.ts
import bcrypt from 'bcryptjs';
import jwt from 'jsonwebtoken';
import { DbUser } from '../models/db.js';
import { UserRepository } from '../repositories/UserRepository.js';

export interface AuthTokens {
  accessToken: string;
  refreshToken: string;
  user: {
    id: string;
    email: string;
    name: string;
    role: string;
  };
}

export class AuthService {
  private static readonly JWT_SECRET = process.env.JWT_SECRET || 'feedbackpro_dev_secret_key_32_chars_min';
  private static readonly ACCESS_TOKEN_EXPIRY = '15m';

  public static async register(email: string, password: string, name: string): Promise<AuthTokens> {
    const existing = await UserRepository.findByEmail(email);
    if (existing) {
      throw new Error('An account with this email already exists');
    }

    if (password.length < 8) {
      throw new Error('Password must be at least 8 characters long');
    }

    const salt = await bcrypt.genSalt(12);
    const passwordHash = await bcrypt.hash(password, salt);

    const user = await UserRepository.create(email, passwordHash, name);
    return await this.generateTokenPair(user);
  }

  public static async login(email: string, password: string): Promise<AuthTokens> {
    const user = await UserRepository.findByEmail(email);
    if (!user) {
      throw new Error('Invalid email or password');
    }

    const isMatch = await bcrypt.compare(password, user.passwordHash);
    if (!isMatch) {
      throw new Error('Invalid email or password');
    }

    return await this.generateTokenPair(user);
  }

  /**
   * Refresh Token Rotation with Reuse / Replay Detection.
   */
  public static async refreshAccessToken(oldRefreshToken: string): Promise<AuthTokens> {
    const record = await UserRepository.findRefreshToken(oldRefreshToken);
    if (!record) {
      throw new Error('Invalid refresh token');
    }

    // Reuse detection: If token was already revoked, someone may have compromised the family!
    if (record.revokedAt) {
      await UserRepository.revokeAllUserTokens(record.userId);
      throw new Error('Compromised refresh token detected. All sessions revoked for security.');
    }

    if (new Date() > record.expiresAt) {
      throw new Error('Refresh token has expired. Please log in again.');
    }

    const user = await UserRepository.findById(record.userId);
    if (!user) {
      throw new Error('User account not found');
    }

    // Issue new pair & rotate
    const newTokens = await this.generateTokenPair(user);
    await UserRepository.rotateRefreshToken(oldRefreshToken, newTokens.refreshToken);

    return newTokens;
  }

  public static async logout(refreshToken?: string): Promise<void> {
    if (refreshToken) {
      const record = await UserRepository.findRefreshToken(refreshToken);
      if (record) {
        await UserRepository.rotateRefreshToken(refreshToken, 'revoked_on_logout');
      }
    }
  }

  public static verifyAccessToken(token: string): { userId: string; role: string } {
    try {
      const decoded = jwt.verify(token, this.JWT_SECRET) as { userId: string; role: string };
      return decoded;
    } catch {
      throw new Error('Invalid or expired authentication token');
    }
  }

  private static async generateTokenPair(user: DbUser): Promise<AuthTokens> {
    const accessToken = jwt.sign(
      { userId: user.id, email: user.email, role: user.role },
      this.JWT_SECRET,
      { expiresIn: this.ACCESS_TOKEN_EXPIRY }
    );

    const refreshToken = `${user.id}.${Date.now()}.${Math.random().toString(36).substring(2, 15)}`;
    await UserRepository.storeRefreshToken(user.id, refreshToken, 7);

    return {
      accessToken,
      refreshToken,
      user: {
        id: user.id,
        email: user.email,
        name: user.name,
        role: user.role
      }
    };
  }
}

File: backend/app/services/auth_service.py
"""Authentication Service with Argon2id and Refresh Token Rotation/Reuse Detection."""

from datetime import datetime, timedelta, timezone
import uuid

from fastapi import HTTPException, status
from sqlalchemy import select, update
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import get_settings
from app.core.security import (
    create_access_token,
    generate_secure_token,
    hash_password,
    hash_token,
    verify_password,
)
from app.db.models.user import RefreshToken, User
from app.schemas.auth import TokenResponse, UserLoginRequest, UserRegisterRequest

settings = get_settings()


class AuthService:
    @classmethod
    async def register_user(cls, db: AsyncSession, data: UserRegisterRequest) -> User:
        clean_email = data.email.lower().strip()

        # Check existing user
        existing = await db.scalar(select(User).where(User.email == clean_email))
        if existing:
            raise HTTPException(
                status_code=status.HTTP_400_BAD_REQUEST,
                detail="An account with this email address already exists.",
            )

        hashed_pw = hash_password(data.password)
        new_user = User(
            email=clean_email,
            password_hash=hashed_pw,
            name=data.name.strip(),
            role="user",
            email_verified=False,
        )
        db.add(new_user)
        await db.commit()
        await db.refresh(new_user)
        return new_user

    @classmethod
    async def login_user(cls, db: AsyncSession, data: UserLoginRequest) -> TokenResponse:
        clean_email = data.email.lower().strip()
        user = await db.scalar(select(User).where(User.email == clean_email))
        if not user or not verify_password(data.password, user.password_hash):
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid email or password.",
                headers={"WWW-Authenticate": "Bearer"},
            )

        # Generate tokens
        access_token = create_access_token(user_id=user.id, role=user.role)
        plain_refresh_token = generate_secure_token()
        token_hash_val = hash_token(plain_refresh_token)

        now = datetime.now(timezone.utc)
        expires_at = now + timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS)

        refresh_entry = RefreshToken(
            token_hash=token_hash_val,
            user_id=user.id,
            expires_at=expires_at,
        )
        db.add(refresh_entry)
        await db.commit()

        return TokenResponse(
            access_token=access_token,
            refresh_token=plain_refresh_token,
            token_type="bearer",
            expires_in=settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60,
        )

    @classmethod
    async def rotate_refresh_token(cls, db: AsyncSession, refresh_token_str: str) -> TokenResponse:
        token_hash_val = hash_token(refresh_token_str)
        token_entry = await db.scalar(select(RefreshToken).where(RefreshToken.token_hash == token_hash_val))

        now = datetime.now(timezone.utc)

        if not token_entry:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Invalid or expired refresh token.",
            )

        # REUSE DETECTION: If token is already revoked or replaced, an attacker or compromised client is reusing it
        if token_entry.revoked_at is not None or token_entry.replaced_by_token is not None:
            # Revoke ALL refresh tokens for this user immediately as a safety precaution
            await db.execute(
                update(RefreshToken)
                .where(RefreshToken.user_id == token_entry.user_id)
                .values(revoked_at=now)
            )
            await db.commit()
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Security alert: Refresh token reuse detected. All active sessions have been revoked.",
            )

        # Check expiration
        if token_entry.expires_at.tzinfo is None:
            entry_expires = token_entry.expires_at.replace(tzinfo=timezone.utc)
        else:
            entry_expires = token_entry.expires_at

        if entry_expires < now:
            token_entry.revoked_at = now
            await db.commit()
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Refresh token has expired. Please log in again.",
            )

        # Fetch user
        user = await db.get(User, token_entry.user_id)
        if not user:
            raise HTTPException(status_code=status.HTTP_401_UNAUTHORIZED, detail="User account not found.")

        # Rotate: Mark old token as revoked and generate new pair
        new_plain_refresh = generate_secure_token()
        new_token_hash = hash_token(new_plain_refresh)

        token_entry.revoked_at = now
        token_entry.replaced_by_token = new_token_hash

        new_entry = RefreshToken(
            token_hash=new_token_hash,
            user_id=user.id,
            expires_at=now + timedelta(days=settings.REFRESH_TOKEN_EXPIRE_DAYS),
        )
        db.add(new_entry)

        new_access_token = create_access_token(user_id=user.id, role=user.role)
        await db.commit()

        return TokenResponse(
            access_token=new_access_token,
            refresh_token=new_plain_refresh,
            token_type="bearer",
            expires_in=settings.ACCESS_TOKEN_EXPIRE_MINUTES * 60,
        )

    @classmethod
    async def logout_user(cls, db: AsyncSession, refresh_token_str: str) -> None:
        token_hash_val = hash_token(refresh_token_str)
        token_entry = await db.scalar(select(RefreshToken).where(RefreshToken.token_hash == token_hash_val))
        if token_entry:
            token_entry.revoked_at = datetime.now(timezone.utc)
            await db.commit()
